0015-ilp-addresses/0015-ilp-addresses.md

```diff
@@ -0,0 +1,179 @@
+
+## Routing
+
+Connectors maintain a _routing table_ of ILP addresses. Routing is a recursive lookup through the routing tables of any number of connectors. When a connector receives a query, it finds the [longest prefix match](https://en.wikipedia.org/wiki/Longest_prefix_match) for the queried address. Then, it follows one of the following cases:
+
+- If the matching address is marked for local delivery, the connector prepares a transfer to that address in one of the ledgers connected to it. The connector maps the ILP address to an account within the ledger. (This is the base case.)
+- If the matching address is marked as forwarded delivery, it has the address of another connector associated with it in the routing table. The connector makes a routing lookup on the connector associated with the address. (This is the recursive case.)
+
+## Address Requirements
+
+ILP Addresses must meet the following requirements:
+
+1. The address MUST begin with a prefix indicating the allocation scheme. See [Allocation Schemes](#allocation-schemes) for more information.
+2. Each "segment" of the address MUST contain one or more of the following characters:
+    - Alphanumeric characters, upper or lower case. (Addresses are **case-sensitive** so that they can contain data encoded in formats such as base64url.)
+    - Underscore (`_`)
+    - Tilde (`~`)
+    - Hyphen (`-`)
+3. Each segment MUST be separated from other segments by a period character (`.`).
+4. Address prefixes MUST end in a period (`.`) character and MAY contain any number of segments after the allocation scheme prefix.
+5. Destination addresses MUST NOT end in a period (`.`) character, and MUST contain at least two segments after the allocation scheme prefix.
+6. The total length of an ILP Address must be no more than **1023 characters** including the allocation scheme prefix, separators, and all segments.
+
+The following ABNF specification defines the format for all ILP addresses and address prefixes:
+
+```abnf
+address     = scheme separator *prefix [segment]
+                    ; the last segment is REQUIRED for destination addresses
+
+scheme      = "g" / "private" / "example" / "peer" / "self" /
+              "test1" / "test2" / "test3"
+
+separator   = "."
+
+prefix      = 1*(segment separator)
+
+segment     = 1*( ALPHA / DIGIT / "_" / "~" / "-" )
+```
+
+You can also use the following regular expressions to verify the same requirements:
+
+| Address Type        | Regular Expression                                     |
+|:--------------------|:-------------------------------------------------------|
+| All addresses       | `^(g|private|example|peer|self|test[1-3])\.([a-zA-Z0-9_~-]+\.)*([a-zA-Z0-9_~-]+)?$` |
+| Address prefix      | `^(g|private|example|peer|self|test[1-3])\.([a-zA-Z0-9_~-]+\.)*$` |
+| Destination address | `^(g|private|example|peer|self|test[1-3])\.([a-zA-Z0-9_~-]+\.)+[a-zA-Z0-9_~-]+$` |
```

Comment:
I propose we change in all three regular expressions the part matching a dot from `\.` to `[.]`. The resulting regular expression is [equivalent](https://regex101.com/r/K3hnzg/2) in terms of matching the same strings, but it is less prone to copy-paste errors. 

In most programming languages, the backslash (`\`) is used as the escape character in string literals. Declaring a string literal with one of the regular expressions from above would lead to the `\.` being interpreted as an escape sequence and the `\` would be removed from the string. In consequence, the regular expression does not match the expected address format. For example:

```javascript
// javascript
const allAddresses = `^(g|private|example|peer|self|test[1-3])\.([a-zA-Z0-9_~-]+\.)*([a-zA-Z0-9_~-]+)?$`
const re = new RegExp(allAddresses)
console.log(re) // /^(g|private|example|peer|self|test[1-3]).([a-zA-Z0-9_~-]+.)*([a-zA-Z0-9_~-]+)?$/
```

With the proposed change, we would avoid the backslash being interpreted as an escape character in the string literal:

```javascript
// javascript
const allAddresses = `^(g|private|example|peer|self|test[1-3])[.]([a-zA-Z0-9_~-]+[.])*([a-zA-Z0-9_~-]+)?$`
const re = new RegExp(allAddresses)
console.log(re) // /^(g|private|example|peer|self|test[1-3])[.]([a-zA-Z0-9_~-]+[.])*([a-zA-Z0-9_~-]+)?$/
```